Python/pylifecycle.c

```diff
@@ -2515,32 +2513,42 @@ _Py_DumpExtensionModules(int fd, PyInterpreterState *interp)
         return;
     }
 
+    Py_ssize_t pos;
+    PyObject *key, *value;
+
+    // Avoid PyDict_GetItemString() which calls PyUnicode_FromString(),
+    // memory cannot be allocated on the heap in a signal handler.
+    // Iterate on the dict instead.
+    PyObject *module_names = NULL;
+    pos = 0;
+    while (PyDict_Next(interp->sysdict, &pos, &key, &value)) {
+        if (PyUnicode_Check(key)
+           && PyUnicode_CompareWithASCIIString(key, "module_names") == 0) {
+            module_names = value;
+            break;
+        }
+    }
+    // If we failed to get sys.module_names or it's not a tuple,
+    // don't exclude stdlib modules.
+    if (module_names != NULL && !PyTuple_Check(module_names)) {
+        module_names = NULL;
+    }
+
+    // List extensions
     int header = 1;
     Py_ssize_t count = 0;
-    Py_ssize_t pos = 0;
-    PyObject *key, *value;
+    pos = 0;
     while (PyDict_Next(modules, &pos, &key, &value)) {
         if (!PyUnicode_Check(key)) {
             continue;
         }
         if (!_PyModule_IsExtension(value)) {
             continue;
         }
-
-        // Check if it is a stdlib extension module.
         // Use the module name from the sys.modules key,
         // don't attempt to get the module object name.
-        const Py_ssize_t names_len = Py_ARRAY_LENGTH(_Py_module_names);
-        int is_stdlib_mod = 0;
-        for (Py_ssize_t i=0; i < names_len; i++) {
-            const char *name = _Py_module_names[i];
-            if (PyUnicode_CompareWithASCIIString(key, name) == 0) {
-                is_stdlib_mod = 1;
-                break;
-            }
-        }
-        if (is_stdlib_mod) {
-            // Ignore stdlib extension module.
+        if (module_names != NULL && PySequence_Contains(module_names, key)) {
```

Comment:
`PySequence_Contains` can raise an exception, no?

Reply:
Since module_names can only be a tuple at this point, the code calls tuplecontains():

```
    for (i = 0, cmp = 0 ; cmp == 0 && i < Py_SIZE(a); ++i)
        cmp = PyObject_RichCompareBool(PyTuple_GET_ITEM(a, i), el, Py_EQ);
```

Well, I should inline the code and call PyUnicode_Compare() directly, to prevent any bad surprises with sys.module_names which contains something different than a string.

Reply:
> Well, I should inline the code and call PyUnicode_Compare() directly, to prevent any bad surprises with sys.module_names which contains something different than a string.

👍